go/tools/builders/vet.go

```diff
@@ -0,0 +1,184 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// vet runs the 'go vet tool' command with the given vet configuration file and
+// returns the emitted findings. It returns an error only if vet emits no
+// findings and at least one error message.
+func runVet(vetTool, vcfg string) (string, error) {
+	args := append(vetFlags, vcfg)
+	cmd := exec.Command(vetTool, args...)
+	out, err := cmd.CombinedOutput()
+	// Note: 'go tool vet' emits a non-zero return code both when vet encounters
+	// an internal error and when vet finds legitimate issues.
+	if err == nil {
+		return "", nil
+	}
+	vetFindings, vetErrors := []string{}, []string{}
+	errMsgs := splitOutput(string(out))
+	for _, m := range errMsgs {
+		if !vetErrorMsgPattern.MatchString(m) {
+			vetErrors = append(vetErrors, m)
+			continue
+		}
+		vetFindings = append(vetFindings, m)
+	}
+	if len(vetFindings) == 0 && len(vetErrors) != 0 {
+		return "", errors.New(strings.Join(vetErrors, "\n"))
+	}
+	return strings.Join(vetFindings, "\n"), nil
+}
+
+var vetFlags = []string{
+	// NOTE: Keep in sync with github.com/golang/go/src/cmd/go/internal/test/test.go
+	"-atomic",
+	"-bool",
+	"-buildtags",
+	"-nilfunc",
+	"-printf",
+}
+
+// vetErrorMsgPattern matches an error message emitted by vet.
+//
+// This regexp should be strict enough to exclude internal errors (prefixed with
+// "vet") and type-check errors (which additionally include a column number).
+//
+// NOTE: this might break if the formating of vet error messages changes.
+var vetErrorMsgPattern = regexp.MustCompile(`^([^:]+?):([0-9]+): (.*)`)
+
+// splitOutput was adapted from go/test/run.go.
+func splitOutput(out string) []string {
+	// gc error messages continue onto additional lines with leading tabs.
+	// Split the output at the beginning of each line that doesn't begin with a tab.
+	// <autogenerated> lines are impossible to match so those are filtered out.
+	var res []string
+	for _, line := range strings.Split(out, "\n") {
+		line = strings.TrimSuffix(line, "\r") // normalize Windows output
+		if strings.HasPrefix(line, "\t") {
+			res[len(res)-1] += "\n" + line
+		} else if strings.HasPrefix(line, "go tool") || strings.HasPrefix(line, "#") {
+			continue
+		} else if strings.TrimSpace(line) != "" {
+			res = append(res, line)
+		}
+	}
+	return res
+}
+
+// parseArchivesAndImportMap parses the -archivefile and -importmap arguments
+// created by the compile rule into the packageFiles and importMap maps used
+// by the vet and checker importers.
+func parseArchivesAndImportMap(archiveFiles, importMapIn []string) (packageFiles map[string]string, importMap map[string]string, err error) {
```

Comment:
This function should probably be in checker_main.go, since its output is also used for the importer.

Reply:
Done.